File: src/main/java/xyz/phanta/cbtweaker/integration/jei/ingredient/JeiIngredientProviderMap.java
package xyz.phanta.cbtweaker.integration.jei.ingredient;

import xyz.phanta.cbtweaker.buffer.BufferType;

import java.util.*;

public class JeiIngredientProviderMap {

    private final Map<BufferType<?, ?, ?, ?>, List<? extends JeiIngredientProvider<?>>> providerTable = new HashMap<>();

    public <JA> void put(BufferType<?, ?, ?, JA> bufType, List<JeiIngredientProvider<JA>> providers) {
        providerTable.put(bufType, providers);
    }

    @SuppressWarnings("unchecked")
    public <JA> List<JeiIngredientProvider<JA>> get(BufferType<?, ?, ?, JA> bufType) {
        List<? extends JeiIngredientProvider<?>> providers = providerTable.get(bufType);
        return providers != null ? (List<JeiIngredientProvider<JA>>)providers : Collections.emptyList();
    }

    @SuppressWarnings("unchecked")
    public <JA> List<JeiIngredientProvider<JA>> getOrCreate(BufferType<?, ?, ?, JA> bufType) {
        return (List<JeiIngredientProvider<JA>>)providerTable.computeIfAbsent(bufType, k -> new ArrayList<>());
    }

    public Collection<BufferType<?, ?, ?, ?>> getBufferTypes() {
        return providerTable.keySet();
    }

    public boolean forEach(Visitor visitor) {
        for (Map.Entry<BufferType<?, ?, ?, ?>, List<? extends JeiIngredientProvider<?>>> entry
                : providerTable.entrySet()) {
            if (!visitUnchecked(visitor, entry.getKey(), entry.getValue())) {
                return false;
            }
        }
        return true;
    }

    @SuppressWarnings("unchecked")
    private static <JB, JA> boolean visitUnchecked(Visitor visitor, BufferType<?, ?, JB, JA> bufType,
                                                   List<? extends JeiIngredientProvider<?>> providers) {
        return visitor.visit(bufType, (List<JeiIngredientProvider<JA>>)providers);
    }

    public interface Visitor {

        <JB, JA> boolean visit(BufferType<?, ?, JB, JA> bufType, List<JeiIngredientProvider<JA>> providers);

    }

}

File: src/main/kotlin/st/evening/mc/cbtweaker/compat/jei/ingredient/impl/JeiIngredientProviderMap.kt
package st.evening.mc.cbtweaker.compat.jei.ingredient.impl

import st.evening.mc.cbtweaker.buffer.BufferType
import st.evening.mc.cbtweaker.compat.jei.ingredient.JeiIngredientProvider

class JeiIngredientProviderMap {
    private val providerTable: MutableMap<BufferType<*, *, *, *>, List<JeiIngredientProvider<*>>> =
        mutableMapOf()

    val bufferTypes: Collection<BufferType<*, *, *, *>>
        get() = providerTable.keys

    operator fun <JA> set(bufType: BufferType<*, *, *, JA>, providers: List<JeiIngredientProvider<JA>>) {
        providerTable[bufType] = providers
    }

    @Suppress("UNCHECKED_CAST")
    operator fun <JA> get(bufType: BufferType<*, *, *, JA>): List<JeiIngredientProvider<JA>> =
        providerTable[bufType] as? List<JeiIngredientProvider<JA>> ?: emptyList()

    fun forEach(visitor: Visitor): Boolean {
        providerTable.entries.forEach { (bufType, providers) ->
            @Suppress("UNCHECKED_CAST")
            if (!visitor.visit(bufType as BufferType<*, *, *, Any>, providers as List<JeiIngredientProvider<Any>>)) {
                return false
            }
        }
        return true
    }

    interface Visitor {
        fun <JB, JA> visit(bufType: BufferType<*, *, JB, JA>, providers: List<JeiIngredientProvider<JA>>): Boolean
    }
}
